### PR TITLE
Terminate DBSCAN's neighbor counting early once threshold is reached

### PR DESCRIPTION
**Pull request by aprokop**

### CODE REVIEW

File: examples/dbscan/ArborX_DBSCAN.hpp [hunk 1]
```diff
@@ -52,15 +52,23 @@ namespace DBSCAN
 {
 
 template <typename MemorySpace>
-struct NumNeighCallback
+struct NumNeighEarlyExitCallback
```
Comment (dalg24) [inline examples/dbscan/ArborX_DBSCAN.hpp:55]: Nitpick: no need to rename the callback.

**Review by dalg24 — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

File: examples/dbscan/ArborX_DBSCAN.hpp (hunk 1 above)
Comment (aprokop) [reply examples/dbscan/ArborX_DBSCAN.hpp:55]: I just thought that if the original name is retained, I would expect the result to be the exact number of neighbors, which is no longer true.

**Comment by aprokop** [comment]
Some results on V100 using HACC data (ran as `./ArborX_DBSCAN.exe --cluster-min-size=100 --binary --eps 0.042 --filename hacc_points.dat --print-dbscan-timers --core-min-size \<core_size\>`

| core_min_size | DBSCAN (master) | DBSCAN (PR) | Speedup |
| ------------- | --------------- | ----------- | ------- |
| 2 | 1.53 | 1.16 | 1.31 |
| 5 | 1.53 | 1.17 | 1.30 |
| 10 | 1.52 | 1.18 | 1.28 |
| 50 | 1.49 | 1.23 | 1.21 |
| 100 | 1.45 | 1.27 | 1.14 |
| 250 | 1.38 | 1.28 | 1.07 |
| 2000 | 1.00 | 1.03 | 0.97 |

Obviously, as `core_min_size` increases, exiting early becomes less efficient.

**Comment by aprokop** [comment]
Merging with HIP failure:
```
/var/jenkins/workspace/ArborX_PR-428/test/tstLinearBVH.cpp(917): error: in "structured_grid<Kokkos__Device<Kokkos__Experimental__HIP_ Kokkos__Experimental__HIPSpace>>": check indices_host(i) == i has failed [702 != 701]
```

**Pull request merged**